### PR TITLE
Fix mypy typing errors in pytorch_lightning/trainer/trainer.py

### PR DESCRIPTION
**Pull request by BongYang**
## What does this PR do?
Part of https://github.com/Lightning-AI/lightning/issues/13445.

Fix typing errors in pytorch_lightning/trainer/trainer.py

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: src/pytorch_lightning/trainer/trainer.py [hunk 1]
```diff
@@ -154,7 +154,7 @@ def __init__(
         fast_dev_run: Union[int, bool] = False,
         accumulate_grad_batches: Optional[Union[int, Dict[int, int]]] = None,
         max_epochs: Optional[int] = None,
-        min_epochs: Optional[int] = None,
+        min_epochs: int = 0,
```
Comment (BongYang) [inline src/pytorch_lightning/trainer/trainer.py:157]: I don't know why min_epochs has None as default.
I think it should be zero as default

but I found some test cases that min_epochs is None.
https://github.com/Lightning-AI/lightning/blob/deaadc157b87b653f1c3aa7926a5ff092ab81863/tests/tests_pytorch/trainer/flags/test_min_max_epochs.py#L21
https://github.com/Lightning-AI/lightning/blob/deaadc157b87b653f1c3aa7926a5ff092ab81863/tests/tests_pytorch/callbacks/test_timer.py#L134
Should I put it back?

**Comment by Borda** [comment]
@BongYang, great work; mind checking the collisions? :rabbit:

**Comment by BongYang** [comment]
> @BongYang, great work; mind checking the collisions? 🐰

@Borda Done

but there is import error for get_len
how can I solve this?

**Comment by otaj** [comment]
Hi @BongYang, the issue was with circular imports. I took the liberty to resolve the import errors for you :zap:

File: src/pytorch_lightning/trainer/trainer.py [hunk 2]
```diff
@@ -2483,10 +2498,11 @@ def current_epoch(self) -> int:
 
     @property
     def max_epochs(self) -> int:
+        assert self.fit_loop.max_epochs
```
Comment (otaj) [inline src/pytorch_lightning/trainer/trainer.py:2501]: It seems this line is causing some issues within CI

**Review by otaj — COMMENTED** [review]
Hi @BongYang, thank you, for the PR! I left couple comments. It might seem, like it's a lot of them, but most of them are actually variations on similar ideas, that just happened to be present in multiple places in the PR :zap:

File: src/pytorch_lightning/trainer/trainer.py [hunk 3]
```diff
@@ -965,7 +972,7 @@ def _predict_impl(
         self.state.status = TrainerStatus.RUNNING
         self.predicting = True
 
-        self.predict_loop.return_predictions = return_predictions
+        self.predict_loop.return_predictions = return_predictions  # type: ignore
```
[the remark this answers is not in the document]
Comment (BongYang) [reply src/pytorch_lightning/trainer/trainer.py]: fix https://github.com/Lightning-AI/lightning/pull/14204/commits/7f62481aaac83f9bdebc0a7e05546ec68f1e57c6

File: src/pytorch_lightning/trainer/trainer.py [hunk 4]
```diff
@@ -721,7 +728,7 @@ def _fit_impl(
         # TODO: ckpt_path only in v2.0
         ckpt_path = ckpt_path or self.resume_from_checkpoint
         self._ckpt_path = self.__set_ckpt_path(
-            ckpt_path, model_provided=True, model_connected=self.lightning_module is not None
+            ckpt_path, model_provided=True, model_connected=self.lightning_module is not None  # type: ignore
```
[the remark this answers is not in the document]
Comment (BongYang) [reply src/pytorch_lightning/trainer/trainer.py]: fix https://github.com/Lightning-AI/lightning/pull/14204/commits/7f62481aaac83f9bdebc0a7e05546ec68f1e57c6

File: src/pytorch_lightning/strategies/strategy_registry.py [hunk 5]
```diff
@@ -83,7 +83,7 @@ def do_register(strategy: Callable) -> Callable:
 
         return do_register
 
-    def get(self, name: str, default: Optional[Any] = None) -> Any:
+    def get(self, name: str, default: Optional[Strategy] = None) -> Strategy:  # type: ignore
```
[the remark this answers is not in the document]
Comment (BongYang) [reply src/pytorch_lightning/strategies/strategy_registry.py]: fix https://github.com/Lightning-AI/lightning/pull/14204/commits/7f62481aaac83f9bdebc0a7e05546ec68f1e57c6

File: src/pytorch_lightning/trainer/trainer.py [hunk 6]
```diff
@@ -1428,8 +1437,8 @@ def __set_ckpt_path(self, ckpt_path: Optional[str], model_provided: bool, model_
                     f'.{fn}(ckpt_path="last") is set, but there is no fault tolerant'
                     " or last checkpoint available. No checkpoint will be loaded."
                 )
-                return
-            ckpt_path = max(candidates_ts.keys(), key=partial(operator.getitem, candidates_ts))
+                return None
+            ckpt_path = max(candidates_ts.keys(), key=partial(operator.getitem, candidates_ts))  # type: ignore
```
[the remark this answers is not in the document]
Comment (BongYang) [reply src/pytorch_lightning/trainer/trainer.py]: fix https://github.com/Lightning-AI/lightning/pull/14204/commits/7f62481aaac83f9bdebc0a7e05546ec68f1e57c6

File: src/pytorch_lightning/trainer/trainer.py [hunk 7]
```diff
@@ -610,17 +615,19 @@ def _setup_on_init(self) -> None:
...
-        self.predict_dataloaders = None
-        self._last_train_dl_reload_epoch = None
-        self._last_val_dl_reload_epoch: Optional[int] = None
+
+        self.train_dataloader: Optional[Union[CombinedLoader, Any]] = None
+
+        self.num_sanity_val_batches: List[Union[int, float]] = []
+        self.num_test_batches: List[Union[int, float]] = []
+        self.num_val_batches: List[Union[int, float]] = []
+        self.num_predict_batches: List[Union[int, float]] = []
```
[the remark this answers is not in the document]
Comment (BongYang) [reply src/pytorch_lightning/trainer/trainer.py]: fix https://github.com/Lightning-AI/lightning/pull/14204/commits/b58c0349489484b4172265c05914abcfcf319ea4

File: src/pytorch_lightning/loops/dataloader/evaluation_loop.py [hunk 8]
```diff
@@ -213,7 +213,7 @@ def teardown(self) -> None:
         self._results.cpu()
         self.epoch_loop.teardown()
 
-    def _get_max_batches(self) -> List[int]:
+    def _get_max_batches(self) -> List[Union[int, float]]:
```
[the remark this answers is not in the document]
Comment (BongYang) [reply src/pytorch_lightning/loops/dataloader/evaluation_loop.py]: fix https://github.com/Lightning-AI/lightning/pull/14204/commits/b58c0349489484b4172265c05914abcfcf319ea4

File: src/pytorch_lightning/loops/dataloader/prediction_loop.py [hunk 9]
```diff
@@ -52,7 +52,7 @@ def num_dataloaders(self) -> int:
         return length
 
     @property
-    def max_batches(self) -> List[int]:
+    def max_batches(self) -> List[Union[int, float]]:
```
[the remark this answers is not in the document]
Comment (BongYang) [reply src/pytorch_lightning/loops/dataloader/prediction_loop.py]: fix https://github.com/Lightning-AI/lightning/pull/14204/commits/b58c0349489484b4172265c05914abcfcf319ea4

File: src/pytorch_lightning/trainer/trainer.py [hunk 10]
```diff
@@ -610,17 +615,19 @@ def _setup_on_init(self) -> None:
...
-        self.num_test_batches = []
-        self.num_val_batches = []
-        self.num_predict_batches = []
-        self.test_dataloaders = None
-        self.val_dataloaders = None
-        self.predict_dataloaders = None
-        self._last_train_dl_reload_epoch = None
-        self._last_val_dl_reload_epoch: Optional[int] = None
+
+        self.train_dataloader: Optional[Union[CombinedLoader, Any]] = None
```
[the remark this answers is not in the document]
Comment (BongYang) [reply src/pytorch_lightning/trainer/trainer.py]: fix https://github.com/Lightning-AI/lightning/pull/14204/commits/ad676af0cdff139b431e94c8e5d29dbb8a2e962c

File: src/pytorch_lightning/loops/utilities.py [hunk 11]
```diff
@@ -216,7 +221,7 @@ def _reset_progress(loop: Loop) -> None:
             _reset_progress(v)
 
 
-def _set_sampler_epoch(dataloader: DataLoader, epoch: int) -> None:
+def _set_sampler_epoch(dataloader: Union[DataLoader, CombinedLoader], epoch: int) -> None:
```
[the remark this answers is not in the document]
Comment (BongYang) [reply src/pytorch_lightning/loops/utilities.py:224]: fix https://github.com/Lightning-AI/lightning/pull/14204/commits/ad676af0cdff139b431e94c8e5d29dbb8a2e962c

File: src/pytorch_lightning/trainer/trainer.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (otaj) [reply src/pytorch_lightning/trainer/trainer.py:621]: Hi, @BongYang, I feel really bad. I went again over the whole PR and the codebase and of course, I was able to produce `Union[int, float]` in there as you had it there originally. :facepalm: 

I went over your PR again, solved some merge conflicts arising from us moving things around and now it looks in a good condition :+1:

File: src/pytorch_lightning/loops/utilities.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (otaj) [reply src/pytorch_lightning/loops/utilities.py:224]: I also took liberty of resolving this. :zap:

**Review by otaj — APPROVED** [review]

File: src/pytorch_lightning/trainer/connectors/data_connector.py [hunk 12]
```diff
@@ -398,6 +398,7 @@ def _reset_eval_dataloader(
                 )
 
                 if orig_num_batches == 0:
+                    assert isinstance(orig_num_batches, int)
```
Comment (awaelchli) [inline src/pytorch_lightning/trainer/connectors/data_connector.py:401]: wow, I'm surprised this is needed

File: src/pytorch_lightning/tuner/batch_size_scaling.py [hunk 13]
```diff
@@ -277,13 +275,18 @@ def _adjust_batch_size(
...
     if trainer.state.fn == "fit":
+        from pytorch_lightning.trainer.supporters import CombinedLoader
+
         if trainer.train_dataloader is None:
             trainer.reset_train_dataloader()
 
         assert trainer.train_dataloader is not None
         # TODO: should we check val_dataloaders here too?
+
+        assert isinstance(trainer.train_dataloader, CombinedLoader)
```
Comment (awaelchli) [inline src/pytorch_lightning/tuner/batch_size_scaling.py:286]: maybe the assert above can be deleted then?

File: src/pytorch_lightning/trainer/trainer.py [hunk 14]
```diff
@@ -579,7 +586,7 @@ def _fit_impl(
         val_dataloaders: Optional[EVAL_DATALOADERS] = None,
         datamodule: Optional[LightningDataModule] = None,
         ckpt_path: Optional[str] = None,
-    ) -> None:
+    ) -> Optional[Union[_PREDICT_OUTPUT, _EVALUATE_OUTPUT]]:
```
Comment (awaelchli) [inline src/pytorch_lightning/trainer/trainer.py:589]: Is this correct? Since this is fit_impl

File: src/pytorch_lightning/trainer/trainer.py [hunk 15]
```diff
@@ -527,17 +532,19 @@ def _setup_on_init(self) -> None:
...
+        self.num_sanity_val_batches: List[Union[int, float]] = []
+        self.num_test_batches: List[Union[int, float]] = []
+        self.num_val_batches: List[Union[int, float]] = []
+        self.num_predict_batches: List[Union[int, float]] = []
+
+        self.test_dataloaders: Optional[List[DataLoader]] = None
+        self.val_dataloaders: Optional[List[DataLoader]] = None
+        self.predict_dataloaders: Optional[List[DataLoader]] = None
+        self._last_train_dl_reload_epoch = float("-inf")
+        self._last_val_dl_reload_epoch = float("-inf")
```
Comment (awaelchli) [inline src/pytorch_lightning/trainer/trainer.py:547]: In data_connector there is a check `self.trainer._last_train_dl_reload_epoch is None`, wouldn't that one always be false now? can be deleted probably?

similarly, also for the _last_val_dl_reload_epoch

**Review by awaelchli — APPROVED** [review]

File: src/pytorch_lightning/tuner/batch_size_scaling.py (hunk 13 above)
Comment (otaj) [reply src/pytorch_lightning/tuner/batch_size_scaling.py:286]: you're right.

File: src/pytorch_lightning/trainer/trainer.py (hunk 15 above)
Comment (otaj) [reply src/pytorch_lightning/trainer/trainer.py:547]: Good catch!
